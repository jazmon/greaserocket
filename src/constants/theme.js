// @flow
import color from 'color';

const theme = {
  DARK_PRIMARY_COLOR: '#388E3C',
  DEFAULT_PRIMARY_COLOR: '#4CAF50',
  LIGHT_PRIMARY_COLOR: '#C8E6C9',
  TEXT_PRIMARY_COLOR: '#FFFFFF',
  ACCENT_COLOR: '#FFC107',
  PRIMARY_TEXT_COLOR: 'rgba(0, 0, 0, 0.87)',
  SECONDAY_TEXT_COLOR: 'rgba(0, 0, 0, 0.54)',
  DIVIDER_COLOR: 'rgba(0, 0, 0, 0.12)',
  DISABLED_TEXT_COLOR: 'rgba(0, 0, 0, 0.38)',
  WHITE: '#FFF',
  BLACK: '#000',
};

export default {
  ...theme,
  LIGHT_STATUS_BAR: color(theme.DARK_PRIMARY_COLOR).dark(),
};
