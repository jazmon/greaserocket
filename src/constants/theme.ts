// @flow
import color from 'color';

const theme = {
  colors: {
    primary: {
      normal: '#4CAF50',
      dark: '#388E3C',
      light: '#C8E6C9',
    },
    accent: {
      normal: '#FFC107',
    },
    white: '#FFFFFF',
    black: '#000000',
    transparent: 'transparent',
  },
  text: {
    colors: {
      light: '#FFFFFF',
      primary: 'rgba(0, 0, 0, 0.87)',
      secondary: 'rgba(0, 0, 0, 0.54)',
      disabled: 'rgba(0, 0, 0, 0.38)',
      divider: 'rgba(0, 0, 0, 0.12)',
    },
    size: 12,
  },
};

export type ThemeType = typeof theme;

export function getStatusBarTextColor(
  backgroundColor: string = 'light-content',
): 'light-content' | 'dark-content' {
  return color(backgroundColor).dark() ? 'light-content' : 'dark-content';
}

export default theme;
