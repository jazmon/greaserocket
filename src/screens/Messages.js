// @flow
import React, { Component } from 'react';
import { View, Text, TextInput, Button, FlatList } from 'react-native';
import styled from 'styled-components/native';
import SocketIOClient from 'socket.io-client';
import { connect } from 'react-redux';
import ChatMessage from 'components/ChatMessage.tsx';

import type { ReduxState } from 'redux/modules';

type User = {
  created_at: string,
  updated_at: string,
  name: string,
  email: string,
  picture: string,
  nickname: string,
  user_id: string
};

type Message = {
  id: string,
  content: string,
  created_at: string,
  updated_at: string,
  user: User
};

type Props = {
  loading: boolean,
  profile: ?Auth0Profile
};

type State = {
  messages: Array<Message>,
  text: string,
  initialMessages: boolean
};

const Container = styled.View`
  flex: 1;
  flex-direction: column;
  background-color: ${({ theme }) => theme.colors.white};
  margin: 8;
  padding: 8;
`;
const BaseText = styled.Text`
  color: ${({ theme }) => theme.text.colors.primary};
  font-size: 16;
`;

// const ChatMessage = styled.Text`
//   text-align: left;
//   font-size: 12;
// `;

const List = styled.FlatList`


`;

const InputContainer = styled.View`
  justify-content: flex-end;
  display: flex;
  margin-top: 8;
`;

const MessageArea = styled.View`
  flex-grow: 1;
`;

const extractKey = ({ id }: Message) => id;

class Messages extends Component<*, Props, State> {
  socket: Object;
  list: Object;

  static defaultProps = {
    profile: null,
  };
  state = {
    messages: [],
    text: '',
    initialMessages: false,
  };

  componentDidMount() {
    this.socket = SocketIOClient('http://localhost:9000');
    this.socket.on('message_emitted', this.onReceiveMessage);
    fetch('http://localhost:9000/messages')
      .then(res => res.json())
      .then((json: { data: Array<Message>, error: boolean }) => {
        console.log('json', json);
        this.setState({ messages: json.data });
      })
      .catch(err => alert(err.message) || console.error(err.message, err));
    if (this.props.profile) {
      this.onShouldConnect(this.props.profile);
    }
    if (this.props.messages) {
      this.list.scrollToEnd();
    }
  }

  componentWillReceiveProps(nextProps: Props) {
    if (!this.props.profile && nextProps.profile) {
      this.onShouldConnect(nextProps.profile);
    }

    if (nextProps.messages && !this.state.initialMessages) {
      this.setState({ initialMessages: true }, () => this.list.scrollToEnd());
    }
  }

  componentWillUnmount() {
    if (this.props.profile) {
      this.socket.emit('disconnect');
    }
  }

  bindList = (list: Object) => {
    this.list = list;
  };

  onShouldConnect = (profile: Auth0Profile) => {
    this.socket.emit('connection', {
      userId: profile.userId,
      email: profile.email,
      name: profile.name,
      picture: profile.picture,
      nickname: profile.nickname,
    });
  };

  // onReceiveMessage = (messages: Array<string>) => {
  onReceiveMessage = (message: Message) => {
    this.setState(
      (prevState: State) => ({
        messages: [...prevState.messages, message],
      }),
      () => this.list.scrollToEnd()
    );
  };

  sendMessage = () => {
    if (!this.props.profile || this.state.text === '') return;
    this.socket.emit('new_message', {
      content: this.state.text,
      userId: this.props.profile.userId,
    });
    this.setState({
      text: '',
    });
  };

  render() {
    const { messages, text } = this.state;
    console.log('this.state.messages', this.state.messages);
    return (
      <Container>
        <MessageArea>
          <List
            ref={this.bindList}
            data={messages}
            renderItem={({ item: message }: { item: Message }) => (
              <ChatMessage key={message.id} message={message} />
            )}
            keyExtractor={extractKey}
          />
        </MessageArea>
        <InputContainer>
          <TextInput
            style={{ height: 40, borderColor: 'gray', borderWidth: 1 }}
            onChangeText={txt => this.setState({ text: txt })}
            value={text}
            onSubmitEditing={this.sendMessage}
            returnKeyType="send"
          />
          <Button onPress={this.sendMessage} title="Send" />
        </InputContainer>
      </Container>
    );
  }
}

const mapState = ({ user }: ReduxState) => ({
  ...user,
});

const mapDispatchtoProps = (dispatch: Dispatch) => ({});
export const MessagesComponent = Messages;
export default connect(mapState, mapDispatchtoProps)(Messages);
