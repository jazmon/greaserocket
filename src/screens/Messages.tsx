// @flow
import React, { Component } from 'react';
import { View, Text, TextInput, Button, FlatList } from 'react-native';
import styled from 'styled-components/native';
import SocketIOClient from 'socket.io-client';
import { connect } from 'react-redux';
import ChatMessage from 'components/ChatMessage';
import { Dispatch } from 'redux';
import { Auth0Profile } from 'types/auth';
import { Maybe } from 'types';

import { ReduxState } from 'redux/modules';

type User = {
  created_at: string;
  updated_at: string;
  name: string;
  email: string;
  picture: string;
  nickname: string;
  userId: string;
};

type Message = {
  id: string;
  content: string;
  created_at: string;
  updated_at: string;
  user: User;
};

type Props = {
  loading: boolean;
  profile: Maybe<Auth0Profile>;
};

type State = {
  messages: Message[];
  text: string;
  initialMessages: boolean;
};

const Container = styled.View`
  flex: 1;
  flex-direction: column;
  background-color: ${({ theme }) => theme.colors.white};
  margin: 8px;
  padding: 8px;
`;
const BaseText = styled.Text`
  color: ${({ theme }) => theme.text.colors.primary};
  font-size: 16px;
`;
type Foo = React.ComponentClass<any>;
const List: React.ComponentClass<any> = styled(FlatList)``;

const InputContainer = styled.View`
  justify-content: flex-end;
  display: flex;
  margin-top: 8px;
  flex-direction: row;
`;
const MessageInput = styled.TextInput`
  height: 40px;
  border: 1px solid gray;
  flex-grow: 1;
`;

const MessageArea = styled.View`flex-grow: 1;`;

const extractKey = ({ id }: Message): string => id;

class Messages extends Component<Props, State> {
  socket: {
    emit: (type: string, props?: any) => void;
    on: (type: string, handler: Function) => void;
  };
  list: { scrollToEnd: () => void } | null;

  static defaultProps = {
    profile: null,
  };
  state = {
    messages: [],
    text: '',
    initialMessages: false,
  };

  componentDidMount() {
    this.socket = SocketIOClient('http://localhost:9000');
    this.socket.on('message_emitted', this.onReceiveMessage);
    fetch('http://localhost:9000/v1/messages')
      .then(res => res.json())
      .then((json: { data: Array<Message>; error: boolean }) => {
        console.log('json', json);
        this.setState({ messages: json.data });
      })
      .catch(err => alert(err.message) || console.error(err.message, err));
    if (this.props.profile) {
      this.onShouldConnect(this.props.profile);
    }
    if (this.props.messages) {
      if (this.list) this.list.scrollToEnd();
    }
  }

  componentWillReceiveProps(nextProps: Props) {
    if (!this.props.profile && nextProps.profile) {
      this.onShouldConnect(nextProps.profile);
    }

    if (nextProps.messages && !this.state.initialMessages) {
      this.setState({ initialMessages: true }, () => {
        if (this.list) this.list.scrollToEnd();
      });
    }
  }

  componentWillUnmount() {
    if (this.props.profile) {
      this.socket.emit('disconnect');
    }
  }

  bindList = (list: Object) => {
    this.list = list;
  };

  onShouldConnect = (profile: Auth0Profile) => {
    this.socket.emit('connection', {
      user_id: profile.userId,
      email: profile.email,
      name: profile.name,
      picture: profile.picture,
      nickname: profile.nickname,
    });
  };

  // onReceiveMessage = (messages: Array<string>) => {
  onReceiveMessage = (message: Message) => {
    this.setState(
      (prevState: State) => ({
        messages: [...prevState.messages, message],
      }),
      () => {
        if (this.list) this.list.scrollToEnd();
      },
    );
  };

  sendMessage = () => {
    if (!this.props.profile || this.state.text === '') return;
    this.socket.emit('new_message', {
      content: this.state.text,
      user_id: this.props.profile.userId,
    });
    this.setState({
      text: '',
    });
  };

  render() {
    const { messages, text } = this.state;
    const { profile } = this.props;
    console.log('this.state.messages', this.state.messages);
    return (
      <Container>
        <MessageArea>
          {profile &&
            <List
              ref={this.bindList}
              data={messages}
              renderItem={({ item: message }: { item: Message }) =>
                <ChatMessage
                  key={message.id}
                  message={message}
                  userId={profile.userId}
                />}
              keyExtractor={extractKey}
            />}
        </MessageArea>
        <InputContainer>
          <MessageInput
            onChangeText={txt => this.setState({ text: txt })}
            value={text}
            onSubmitEditing={this.sendMessage}
            returnKeyType="send"
          />
          <Button onPress={this.sendMessage} title="Send" />
        </InputContainer>
      </Container>
    );
  }
}

const mapState = ({ user }: ReduxState) => ({
  ...user,
});

const mapDispatchtoProps = (dispatch: Dispatch<ReduxState>) => ({});
export const MessagesComponent = Messages;
export default connect(mapState, mapDispatchtoProps)(Messages);
