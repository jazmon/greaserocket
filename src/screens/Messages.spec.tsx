import 'react-native';
import * as React from 'react';
// Note: test renderer must be required after react-native.
import * as renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import theme from 'constants/theme';

import { MessagesComponent as Messages } from './Messages';

it('renders correctly', () => {
  const tree = renderer.create(
    <ThemeProvider theme={theme}>
      <Messages />
    </ThemeProvider>,
  );
  expect(tree).toMatchSnapshot();
});
