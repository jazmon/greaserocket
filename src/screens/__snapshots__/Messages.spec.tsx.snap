// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#FFFFFF",
        "flexBasis": 0,
        "flexDirection": "column",
        "flexGrow": 1,
        "flexShrink": 1,
        "marginBottom": 8,
        "marginLeft": 8,
        "marginRight": 8,
        "marginTop": 8,
        "paddingBottom": 8,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 8,
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexGrow": 1,
        },
        undefined,
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "marginTop": 8,
        },
        undefined,
      ]
    }
  >
    <TextInput
      onChangeText={[Function]}
      onSubmitEditing={[Function]}
      returnKeyType="send"
      style={
        Array [
          Object {
            "borderColor": "gray",
            "borderStyle": "solid",
            "borderWidth": 1,
            "flexGrow": 1,
            "height": 40,
          },
          undefined,
        ]
      }
      value=""
    />
    <View
      accessibilityComponentType="button"
      accessibilityLabel={undefined}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      accessible={true}
      collapsable={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          disabled={false}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "padding": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Send
        </Text>
      </View>
    </View>
  </View>
</View>
`;
